packages/nimble-components/src/modal/index.ts

```diff
@@ -0,0 +1,123 @@
+import { attr } from '@microsoft/fast-element';
+import { FoundationElement } from '@microsoft/fast-foundation';
+import { UserDismissed } from '../patterns/dialog/types';
+
+/**
+ * This is a workaround for an incomplete definition of the native dialog element. Remove when using Typescript >=4.8.3.
+ * https://github.com/microsoft/TypeScript/issues/48267
+ * @internal
+ */
+export interface ExtendedDialog extends HTMLDialogElement {
+    showModal(): void;
+    close(): void;
+}
+
+export const ModalState = {
+    opening: 'opening',
+    open: 'open',
+    closing: 'closing',
+    closed: 'closed'
+} as const;
+export type ModalState =
+    (typeof ModalState)[keyof typeof ModalState];
+
+/**
+ * A base class for components based on a native dialog.
+ * @internal
+ */
+// eslint-disable-next-line @typescript-eslint/no-invalid-void-type
+export abstract class Modal<CloseReason = void> extends FoundationElement {
+    /**
+     * @public
+     * @description
+     * Prevents dismissing via the Escape key
+     */
+    @attr({ attribute: 'prevent-dismiss', mode: 'boolean' })
+    public preventDismiss = false;
+
+    /**
+     * The ref to the internal dialog element.
+     *
+     * @internal
+     */
+    public readonly dialogElement!: ExtendedDialog;
+
+    protected state: ModalState = ModalState.closed;
+
+    /**
+     * True if the Modal is open/showing, false otherwise
+     */
+    public get open(): boolean {
+        return this.state === ModalState.open;
```

Comment:
Should the dialog be considered "open" if the state is opening or closing? Do we have test coverage here?

Reply:
It turns out that we do have a test that expects `open` to be true while closing. So I've changed `open` to mean `state !== closed`. Neither Dialog nor Drawer distinguish between `open` and `opening` states, though, so I don't think we need to test that the property `open` is true when `state === opening`.